packages/webdriverio/src/types.ts

```diff
@@ -11,9 +11,78 @@ import type DevtoolsInterception from './utils/interception/devtools'
 import type { Location } from './commands/element/getLocation'
 import type { Size } from './commands/element/getSize'
 
-export type BrowserCommandsType = typeof BrowserCommands
+type $BrowserCommands = typeof BrowserCommands
+type $ElementCommands = typeof ElementCommands
+
+type ElementQueryCommands = '$' | 'custom$' | 'shadow$' | 'react$'
+type ElementsQueryCommands = '$$' | 'custom$$' | 'shadow$$' | 'react$$'
+type ChainablePrototype = {
+    [K in ElementQueryCommands]: (...args: Parameters<$ElementCommands[K]>) => ChainablePromiseElement<ReturnType<$ElementCommands[K]>>
+} & {
+    [K in ElementsQueryCommands]: (...args: Parameters<$ElementCommands[K]>) => ChainablePromiseArray<WebdriverIO.Element>
+}
+
+type AsyncElementProto = {
+    [K in keyof Omit<$ElementCommands, keyof ChainablePrototype>]: OmitThisParameter<$ElementCommands[K]>
+} & ChainablePrototype
+
+interface ChainablePromiseElement<T> extends AsyncElementProto, Promise<T> {
+    /**
+     * WebDriver element reference
+     */
+    elementId: Promise<string>
+    /**
+     * parent of the element if fetched via `$(parent).$(child)`
+     */
+    parent: Promise<WebdriverIO.Element | WebdriverIO.Browser>
+    /**
+     * selector used to fetch this element, can be
+     * - undefined if element was created via `$({ 'element-6066-11e4-a52e-4f735466cecf': 'ELEMENT-1' })`
+     * - a string if `findElement` was used and a reference was found
+     * - or a functin if element was found via e.g. `$(() => document.body)`
+     */
+    selector: Promise<Selector>
+    /**
+     * Error message in case element fetch was not successful
+     */
+    error?: Promise<Error>
+    /**
+     * index of the element if fetched with `$$`
+     */
+    index?: Promise<number>
+}
+interface ChainablePromiseArray<T> extends Promise<T> {
+    /**
+     * Amount of element fetched.
+     */
+    length: Promise<number>
+
+    /**
+     * Unwrap the nth element of the element list.
+     */
+    get: (i: number) => T
```

Comment:
I'd replace `get` method with accessing element by index in the elements array

Reply:
👍  let me take a look